iac-extensions/docker/src/main/java/org/sonar/iac/docker/checks/ConsecutiveRunInstructionCheck.java

```diff
@@ -54,18 +57,49 @@ private static void checkFromRunsInstruction(CheckContext ctx, DockerImage image
   private static List<List<RunInstruction>> extractConsecutiveRunInstructions(DockerImage image) {
     var result = new ArrayList<List<RunInstruction>>();
     var instructions = image.instructions();
-    for (var i = 0; i < instructions.size() - 1; i++) {
+    var i = 0;
+    while (i < instructions.size()) {
       if (instructions.get(i).is(DockerTree.Kind.RUN)) {
-        var subList = new ArrayList<RunInstruction>();
-        while (i < instructions.size() && instructions.get(i).is(DockerTree.Kind.RUN)) {
-          subList.add((RunInstruction) instructions.get(i));
-          i++;
-        }
-        if (subList.size() > 1) {
-          result.add(subList);
-        }
+        i = nextIteration(i, instructions, result);
+      } else {
+        i++;
       }
     }
     return result;
   }
+
+  private static int nextIteration(int i, List<Instruction> instructions, List<List<RunInstruction>> result) {
+    var subList = new ArrayList<RunInstruction>();
+    subList.add((RunInstruction) instructions.get(i));
+    i++;
+    while (i < instructions.size() && instructions.get(i) instanceof RunInstruction current) {
+      var previous = subList.get(subList.size() - 1);
+      if (haveEqualOptions(current, previous)) {
+        subList.add(current);
+        i++;
+      } else {
+        // start next iteration from the current instruction; on the next iteration of the outer loop,
+        // the next instruction will be compared with the current one
+        i--;
+        break;
+      }
+    }
```

Comment:
```suggestion
    RunInstruction currentInstruction = (RunInstruction) instructions.get(i);
    subList.add(currentInstruction);
    while (i + 1 < instructions.size() && instructions.get(i + 1) instanceof RunInstruction nextInstruction) {
      if (haveEqualOptions(currentInstruction, nextInstruction)) {
        subList.add(nextInstruction);
        currentInstruction = nextInstruction;
        i++;
      } else {
        break;
      }
    }
```

I would refactor it like this to simplify the logic, going from a 2-way traversal to a 1-way traversal (only forward, no `get(i - 1)`.

With a 1-way traversal, we could even go further and replace manual `i` index manipulation with an `Iterator`!  
We could replace the `nextIteration`method with a clearer method like `List<RunInstruction> getNextConsecutiveRunInstructions(Iterator<Instruction> instructionIterator)` reducing the number of parameters and potentially making the logic clearer (+ I guess it's nicer to read `instructionIterator.hasNext()`than `i + 1 < instructions.size()` but maybe it's personal preferences 🤔).

I let you decide whether we should go this way or not 😄

Reply:
Great suggestion. I was mostly following the approach that was implemented in the previous version of the rule, so I kept logic with indices